File: java/src/main/java/com/pubnub/api/models/consumer/PNStatus.java
package com.pubnub.api.models.consumer;

import com.pubnub.api.endpoints.remoteaction.RemoteAction;
import com.pubnub.api.enums.PNOperationType;
import com.pubnub.api.enums.PNStatusCategory;
import com.pubnub.core.CoreException;
import com.pubnub.core.Status;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.ToString;
import org.jetbrains.annotations.Nullable;

import java.util.List;

@Builder(toBuilder = true)
@Getter
@ToString
public class PNStatus implements Status {

    private PNStatusCategory category;
    private PNErrorData errorData;
    private boolean error;

    // boolean automaticallyRetry;

    private int statusCode;
    private PNOperationType operation;

    private boolean tlsEnabled;

    @Override
    public Boolean isTlsEnabled() {
        return tlsEnabled;
    }

    private String uuid;
    private String authKey;
    private String origin;
    private Object clientRequest;

    // send back channel, channel groups that were affected by this operation
    @Nullable
    private List<String> affectedChannels;
    @Nullable
    private List<String> affectedChannelGroups;

    @Getter(AccessLevel.NONE)
    @ToString.Exclude
    private RemoteAction<?> executedEndpoint;


    public void retry() {
        executedEndpoint.retry();
    }

    @Nullable
    @Override
    public CoreException getException() {
        if (errorData != null) {
            return (CoreException) errorData.getThrowable(); // TODO figure out way without casting
        }
        return null;
    }

    @Nullable
    @Override
    public Integer __statusCode() {
        return statusCode;
    }

    /*
    public void cancelAutomaticRetry() {
        // TODO
    }
    */

}

File: pubnub-core/pubnub-core-api/src/commonMain/kotlin/com/pubnub/api/models/consumer/PNStatus.kt
package com.pubnub.api.models.consumer

import com.pubnub.api.PubNubException
import com.pubnub.api.enums.PNStatusCategory
import kotlin.jvm.JvmName

class PNStatus(
    val category: PNStatusCategory,
    val exception: PubNubException? = null,
    val currentTimetoken: Long? = null,
    val affectedChannels: Collection<String> = emptySet(),
    val affectedChannelGroups: Collection<String> = emptySet(),
) {
    @get:JvmName("isError")
    val error: Boolean = exception != null

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other !is PNStatus) {
            return false
        }

        if (category != other.category) {
            return false
        }
        if (exception != other.exception) {
            return false
        }
        if (currentTimetoken != other.currentTimetoken) {
            return false
        }
        if (affectedChannels != other.affectedChannels) {
            return false
        }
        if (affectedChannelGroups != other.affectedChannelGroups) {
            return false
        }

        return true
    }

    override fun hashCode(): Int {
        var result = category.hashCode()
        result = 31 * result + (exception?.hashCode() ?: 0)
        result = 31 * result + (currentTimetoken?.hashCode() ?: 0)
        result = 31 * result + affectedChannels.hashCode()
        result = 31 * result + affectedChannelGroups.hashCode()
        return result
    }
}
